src/generic/datavgen.cpp

```diff
@@ -2046,69 +2073,225 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
+        case DropHint_None:
+            break;
+
+        case DropHint_Inside:
+            RefreshRow(m_dropItemInfo.m_row);
+            break;
+
+        case DropHint_Above:
+            RefreshRows(m_dropItemInfo.m_row == 0 ? 0 : m_dropItemInfo.m_row - 1, m_dropItemInfo.m_row);
+            break;
+
+        case DropHint_Below:
+            RefreshRows(m_dropItemInfo.m_row, std::max(m_dropItemInfo.m_row + 1, GetRowCount() - 1));
+            break;
     }
 }
 
-wxDragResult wxDataViewMainWindow::OnDragOver( wxDataFormat format, wxCoord x,
-                                               wxCoord y, wxDragResult def )
+void wxDataViewMainWindow::RemoveDropHint()
+{
+    RefreshDropHint();
+
+    m_dropItemInfo = DropItemInfo();
+}
+
+wxDataViewMainWindow::DropItemInfo wxDataViewMainWindow::GetDropItemInfo(const wxCoord x, const wxCoord y)
 {
+    DropItemInfo dropItemInfo;
+
     int xx = x;
     int yy = y;
     m_owner->CalcUnscrolledPosition( xx, yy, &xx, &yy );
-    unsigned int row = GetLineAt( yy );
 
-    wxDataViewItem item;
+    unsigned int row = GetLineAt(yy);
+    dropItemInfo.m_row = row;
+
+    if (row >= GetRowCount() || xx > GetEndOfLastCol())
+        return dropItemInfo;
+
+    if (IsVirtualList())
+    {
+        dropItemInfo.m_item = GetItemByRow(row);
+
+        if (dropItemInfo.m_item.IsOk())
+            dropItemInfo.m_hint = DropHint_Inside;
+    }
+    else
+    {
+        wxDataViewTreeNode* node = GetTreeNodeByRow(row);
+        if (!node)
+            return dropItemInfo;
+
+        dropItemInfo.m_item = node->GetItem();
+
+        unsigned int itemStart = GetLineStart(row);
+        int itemHeight = GetLineHeight(row);
+
+        bool insertAbove = yy - itemStart < itemHeight*0.15;       // 15%, TODO: configurable
+        if (insertAbove)
+        {
+            // Can be treated as 'insertBelow" with the small difference:
+            node = GetTreeNodeByRow(row - 1);   // We need the node from the previous row
+
+            dropItemInfo.m_hint = DropHint_Above;
+
+            if (!node)
+            {
+                // Seems to be dropped in the root node
+
+                dropItemInfo.m_indentLevel        = 0;
+                dropItemInfo.m_proposedDropIndex  = 0;
+                dropItemInfo.m_item               = wxDataViewItem();
+
+                return dropItemInfo;
+            }
+
+        }
+
+        bool insertBelow = yy - itemStart > itemHeight*0.85;        // 15%, TODO: configurable
+        if (insertBelow)
+            dropItemInfo.m_hint = DropHint_Below;
+
+        if (insertBelow || insertAbove)
+        {
+            // Insert inside the 'item' or after (below) it. Depends on:
+            // 1 - the 'item' is a container
+            // 2 - item's child index in it's parent (the last in the parent or not)
+            // 3 - expanded or not
+            // 4 - mouse x position
+
+            unsigned int start_x = 0;       // Expander column x position start
```

Comment:
```suggestion
            int start_x = 0;       // Expander column x position start
```
We shouldn't use unsigned for anything except counts (and even then...) and making this unsigned results in signed/unsigned comparison warnings below.